wayland-server/src/socket.rs

```diff
@@ -17,20 +16,59 @@ use nix::{
     unistd::unlink,
 };
 
+/// An utility representing a unix socket on which your compositor is listening for new clients
 #[derive(Debug)]
 pub struct ListeningSocket {
     listener: UnixListener,
     _lock: File,
     socket_path: PathBuf,
     lock_path: PathBuf,
-    socket_name: OsString,
+    socket_name: Option<OsString>,
 }
 
 impl ListeningSocket {
+    /// Attempt to bind a listening socket with given name
+    ///
+    /// This method will acquire an associate lockfile. The socket will be created in the
+    /// directory pointed to by the `XDG_RUNTIME_DIR` environment variable.
     pub fn bind<S: AsRef<OsStr>>(socket_name: S) -> Result<Self, BindError> {
         let runtime_dir: PathBuf =
             env::var("XDG_RUNTIME_DIR").map_err(|_| BindError::RuntimeDirNotSet)?.into();
         let socket_path = runtime_dir.join(socket_name.as_ref());
+        let mut socket = Self::bind_absolute(socket_path)?;
+        socket.socket_name = Some(socket_name.as_ref().into());
+        Ok(socket)
+    }
+
+    /// Attempt to bind a listening socket from a sequence of names
+    ///
+    /// This method will repeatedly try to bind sockets in teh form `{basename}-{n}` for values of `n`
+    /// yielded from the provided range and returns the first one that succeeds.
+    ///
+    /// This method will acquire an associate lockfile. The socket will be created in the
+    /// directory pointed to by the `XDG_RUNTIME_DIR` environment variable.
+    pub fn bind_auto(
+        basename: &str,
+        range: impl IntoIterator<Item = usize>,
```

Comment:
Would this be better as a `Range<usize>`?

Reply:
Well, this formulation covers both `Range` and `RangeInclusive`, I didn't find a better way to allow both.